### ISSUE TITLE
Run tests in random order

### ISSUE THREAD

**Issue by AndriusJanusas** [nunit/nunit#663]
Ideally each test should be able to run independently and stand alone, but when you have many of them and they always run in the same order it would be great to run them in random order to ensure, that they're actually independent.

Would it be hard to implement such feature?

**Comment by CharliePoole** [comment]
We have talked about this in the past and it's a little harder than appears at first glance. That's because the "random" order has to be repeatable in order to debug any unexpected dependencies that are discovered. This makes it non-trivial but certainly not impossible. The random ordering would have to be tied in directly to NUnit's own Randomizer, which is used to generate random arguments among other things. In that way, the /seed command-line option or it's equivalent could be used when it came time to debug a particular problem.

[As an aside, we could definitely use some integration-level tests for the repeatability of random values and - if we implement this - test order.]

I'd be inclined to see this as a future enhancement to be added after we implement test ordering itself.

**Comment by rprouse** [comment]
I might even combine this with the planned Order attribute #170. Maybe it could be a property on the Order, something like `[Order(Random=true)]`?

This will prevent the use of random along with an order, but we will need to decide how this will work with other tests that specify an order. Are they random after all ordered tests? Before? Interspersed? Probably after since the order attributes would indicate a requirement for a particular test state that might not be stable if we run random tests first.

**Comment by CharliePoole** [comment]
I guess this depends on how the OrderAttribute works. I saw it as applying to an individual test, not a suite. Randomization OTOH seems like something you would specify on the class affecting all the tests. There are different ways you could handle conflicts but what feels natural to me is having random on the suite simply disable any other ordering at the method level. I can't think of a strong use case for wanting some tests to run randomly and others to run in a strict order.

There's another fundamental issue to resolve about random ordering. Does the random ordering happen when the tests are created or when they are run? If you are running in a future GUI or in the test adapter, should repeated runs re-randomize the tests? Should this be an option? Or is a given order "sticky" once it is generated?

All of this conspires to make me feel this is a bit farther in our future than simple ordering. However, #170 is already marked future, so this one can go the same way without our having to decide how far in the future it is. :-)

**Comment by richseviora** [comment]
How about making random order as a suite/fixture level attribute? That way you'd be able to be explicit about which tests are subject to the randomness, and have that override (or throw an exception) any ordering attributes applied to child tests.

To look at Ruby's RSpec testing framework as an example, the test runner be configured to run examples in random order. It generates a seed number at the time it's run, and the user can supply that in the future to repeat the order. Which naturally requires a modification to any IDEs that support the test runner.

**Comment by CharliePoole** [comment]
@richseviora That's how we are thinking about it. There's already a repeatable seed that is used for generating test cases that use random data. The same seed could be used to generate values for ordering, or we could simply run them in the order of their individual seeds.

**Comment by KthProg** [comment]
What's going on with this?

**Comment by mikkelbu** [comment]
I don't think that anything has happened with this the last 2.5 years. NUnit development is done on voluntary basis in people's spare time, and there are more work than manpower :), so some issues will remain open for a very long time, but PRs are always welcome.

**Comment by Sjlver** [comment]
I think this would be really nice to have. Personally, I'd be happy with an as-simple-as-possible version that doesn't care about reproducibility. Just in case this increases motivation ;-)
